cpp/src/parquet/key_toolkit.h

```diff
@@ -0,0 +1,123 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "parquet/key_encryption_key.h"
+#include "parquet/kms_client.h"
+#include "parquet/kms_client_factory.h"
+#include "parquet/platform.h"
+#include "parquet/two_level_cache_with_expiration.h"
+
+namespace parquet {
+namespace encryption {
+
+class KeyWithMasterID {
```

Comment:
document please.  Does this need to be exported?

Reply:
> Does this need to be exported?

No, it's used internally in FileKeyUnwrapper.cc (implemented) and KeyToolkit.cc (will be implemented later).

Reply:
Can you put internal-only stuff in separate header files? e.g. `key_toolkit_internal.h`?

Reply:
> Can you put internal-only stuff in separate header files?

`KeyWithMasterId` is an internal thing, however, since it appears in public header file `FileKeyUnwrapper.h" (i.e. `KeyWithMasterId FileKeyUnwrapper::GetDataEncryptionKey(const KeyMaterial& key_material) const` - which will be used in KeyToolkit key rotation later). So it must be in a public `key_toolkit.h`.

Reply:
You can use forward declarations to avoid exposing class definitions when only a reference or pointer is given.

Also, it is good idea to hide implementation details by using the pimpl idiom. Another advantage is limit the growth of compile times.